File: 백준/Gold/2643. 색종이 올려 놓기/색종이 올려 놓기.py
# 색종이 올려 놓기
import sys
input = sys.stdin.readline

N = int(input())
papers = []
dp = [0] * N

for i in range(N):
    a, b = map(int,input().split())

    if b > a:
        a, b = b, a
        
    papers.append((a,b))

papers.sort()

# # 작은 종이부터 순회
for i in range(N):
    cur = papers[i][1]
    for j in range(i, -1, -1):
        # 현재 종이위에 쌓을 수 있다면 갱신
        if cur >= papers[j][1]:
            dp[i] = max(dp[i], dp[j]+1)
    
print(max(dp))

File: 백준/Gold/2643. 색종이 올려 놓기/색종이 올려 놓기.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        int N = Integer.parseInt(br.readLine());
        ArrayList<int[]> papers = new ArrayList<>();
        int[] dp = new int[N];
        int result = 0;

        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());

            if (b > a) {
                int tmp = a;
                a = b;
                b = tmp;
            }
            papers.add(new int[]{a, b});
        }

        papers.sort((a, b) -> a[0] != b[0] ? a[0] - b[0] : a[1] - b[1]);

        for (int i = 0; i < N; i++) {
            int cur = papers.get(i)[1];
            for (int j = i; j >= 0; j--) {
                if (cur >= papers.get(j)[1]) {
                    dp[i] = Math.max(dp[i], dp[j] + 1);
                    result = Math.max(result, dp[i]);
                }
            }
        }

        System.out.println(result);
    }
}
